src/main/java/net/snowflake/client/jdbc/SnowflakeFileTransferAgent.java

```diff
@@ -2324,194 +2324,195 @@ private void filterExistingFiles() throws SnowflakeSQLException {
-              if (stream != null) {
-                try {
-                  stream.reset();
-                } catch (IOException ex) {
-                  logger.debug("failed to clean up temp file: {}", ex);
+                for (FileBackedOutputStream stream : fileBackedOutputStreams) {
+                  if (stream != null) {
+                    try {
+                      stream.reset();
+                    } catch (IOException ex) {
+                      logger.debug("failed to clean up temp file: {}", ex);
+                    }
+                  }
                 }
               }
+
+              // continue so that we will upload the file
+              if (hashText == null
+                  || // remote is encrypted & has no digest
+                  (objDigest != null && !hashText.equals(objDigest))
+                  || // digest mismatch
+                  (objDigest == null && !hashText.equals(obj.getMD5()))) // ETag/MD5 mismatch
+              {
+                logger.debug(
+                    "digest diff between remote store and local, will {} {}, "
+                        + "local digest: {}, remote store md5: {}",
+                    commandType.name().toLowerCase(),
+                    mappedSrcFile,
+                    hashText,
+                    obj.getMD5());
+                continue;
+              }
+            } catch (IOException | NoSuchAlgorithmException ex) {
+              throw new SnowflakeSQLLoggedException(
+                  session,
+                  SqlState.INTERNAL_ERROR,
+                  ErrorCode.INTERNAL_ERROR.getMessageCode(),
+                  ex,
+                  "Error reading: " + localFile);
             }
-          }
 
-          // continue so that we will upload the file
-          if (hashText == null
-              || // remote is encrypted & has no digest
-              (objDigest != null && !hashText.equals(objDigest))
-              || // digest mismatch
-              (objDigest == null && !hashText.equals(obj.getMD5()))) // ETag/MD5 mismatch
-          {
             logger.debug(
-                "digest diff between remote store and local, will {} {}, "
-                    + "local digest: {}, remote store md5: {}",
+                "digest same between remote store and local, will not upload {} {}",
                 commandType.name().toLowerCase(),
-                mappedSrcFile,
-                hashText,
-                obj.getMD5());
-            continue;
+                mappedSrcFile);
+
+            skipFile(mappedSrcFile, objFileName);
           }
-        } catch (IOException | NoSuchAlgorithmException ex) {
-          throw new SnowflakeSQLLoggedException(
-              session,
-              SqlState.INTERNAL_ERROR,
-              ErrorCode.INTERNAL_ERROR.getMessageCode(),
-              ex,
-              "Error reading: " + localFile);
+          // exit retry loop
+          break;
```

Comment:
If the for-each loop finishes correctly, we jump out of the do-while retry loop.